4387 video recording for Full Debrief.

saulhidalgoaular opened the pull request:
Fixes #4387

saulhidalgoaular committed:
Partial Integration of the Monte Media library with the Time Controller recording button

saulhidalgoaular committed:
Preference Page in Eclipse.

saulhidalgoaular committed:
Added Enable Video Recording in the preference panel

saulhidalgoaular committed:
Integration with the preference.

saulhidalgoaular committed:
Merge branch 'develop' into 4387_video_recording

saulhidalgoaular committed:
Adding Monte Media as a module of Debrief.

saulhidalgoaular committed:
UI improvements to the Preference menu.

saulhidalgoaular committed:
Integration of Debrief Preference Parameters for Video Recording.

saulhidalgoaular commented:
Uploading ScreenRecording 2023-02-22 at 05.27.32.mov…

saulhidalgoaular commented:
[ScreenRecording 2023-02-22 at 05.27.32.tar.gz](https://github.com/debrief/debrief/files/10801595/ScreenRecording.2023-02-22.at.05.27.32.tar.gz)

saulhidalgoaular committed:
Removing the audio preferences

IanMayo commented:
Hello @saulhidalgoaular - I just downloaded this, and went to play with it. I enabled video recording in the prefs, and recorded the time moving forward. But when I stopped recording the `export to pptx` dialog opened.

Is it a different process to record to video?

That reminds me - we could do with some tutorial content on how to record to video. I'll send you a link.

saulhidalgoaular committed:
Fix null pointer exception due audio support removed.

saulhidalgoaular committed:
Final details to the first release of the Capture Plot to file feature.

IanMayo commented:
Thanks for this @saulhidalgoaular 

The linter has spotted a bunch of improvements. Some are in our code, others in 3rd-party Monte code.  Where the issue is in our code, and is in new code related to this issue, could you investigate/resolve them please?
https://www.codefactor.io/repository/github/debrief/debrief/pull/5167

IanMayo commented on org.mwc.debrief.montemedia/src/MovieMaker.java:
```diff
@@ -0,0 +1,30 @@
+ * @author Werner Randelshofer
+ * @version 1.0 2011-03-19 Created.
+ */
+public class MovieMaker {
```
Hello @saulhidalgoaular  - is there a use-case when we run this app?

saulhidalgoaular replied:
This is a tool to test manually the class that creates .mov files. You can upload some images and an audio and it creates a mov file from that. We don't have an use case in Debrief for this functionality. I am removing it.

By the way, there are many other classes like this one. Would you like me to delete them too? They are some tools to test partial functionality of the library.

IanMayo commented:
Note: on MS-Windows I don't see the drop-down control:
![image](https://user-images.githubusercontent.com/1108513/236449575-efdfbb11-b61c-4c69-a6d8-8a5537023969.png)

MS-Windows also can't replay the recorded video:
![image](https://user-images.githubusercontent.com/1108513/236449748-55fbadeb-449a-4a7d-9fdb-e9004f581b62.png)

IanMayo commented on org.mwc.debrief.montemedia/src/ScreenRecorder.java:
```diff
@@ -0,0 +1,31 @@
+ */
+public class ScreenRecorder {
+
+    /**
```
Aah, @saulhidalgoaular - this is another runtime app. Can you explain what it does please?

saulhidalgoaular replied:
This is part of the original MonteMedia Library. It is a class to use the library without needs of any development. We don't really need it for Debrief. I am removing it.

saulhidalgoaular committed:
Removing extra classes.

saulhidalgoaular committed:
Refactoring and Code Improvement.

saulhidalgoaular committed:
Splitting processClick to reduce warnings.

theanuradha committed:
switch to MP4 format

saulhidalgoaular committed:
Null Pointer Exception when Video Recording is disabled, but it is
chosen from the dropdown. Now It is only shown when It is enabled in the
Preferences section.

saulhidalgoaular commented:
> Note: on MS-Windows I don't see the drop-down control: ![image](https://user-images.githubusercontent.com/1108513/236449575-efdfbb11-b61c-4c69-a6d8-8a5537023969.png)
> 
> MS-Windows also can't replay the recorded video: ![image](https://user-images.githubusercontent.com/1108513/236449748-55fbadeb-449a-4a7d-9fdb-e9004f581b62.png)

Good afternoon. Which Windows' version have you tried? I am running it on Windows 10 and the dropdown button looks like this:

![image](https://github.com/debrief/debrief/assets/11049943/0ee525a6-4019-4a4f-b5e5-976ef21fde56)
![image](https://github.com/debrief/debrief/assets/11049943/32842607-db60-47a1-b297-fae905c8d4d0)

saulhidalgoaular committed:
Revert "switch to MP4 format"

This reverts commit 985463d2565b6d22a2fe58a7a0665ed4e012357b.

saulhidalgoaular committed:
Progress of the integration with ffmpeg. [Not finished]

saulhidalgoaular committed:
FFMPEG integration.

IanMayo commented:
Oh, @saulhidalgoaular and @helenayele - can users choose to only record the Debrief plot window?  That's the most likely recording use-case.

saulhidalgoaular committed:
Merge branch 'develop' into 4387_video_recording

saulhidalgoaular commented:
![image](https://github.com/debrief/debrief/assets/11049943/2799598b-8de1-4be9-8823-f2e1107467ab)

I have added this option. Files generated are playable by Windows Media Player and "Movies & TV" without extra codecs.

saulhidalgoaular committed:
Adding WMV support to the Record to Video feature

saulhidalgoaular commented:
> Oh, @saulhidalgoaular and @helenayele - can users choose to only record the Debrief plot window? That's the most likely recording use-case.

If you mean the only the map, currently it is not supported. I will work on this feature

saulhidalgoaular commented:
> Note: on MS-Windows I don't see the drop-down control: ![image](https://user-images.githubusercontent.com/1108513/236449575-efdfbb11-b61c-4c69-a6d8-8a5537023969.png)

Do you still have the Dropdown the missing?

IanMayo commented:
> > Note: on MS-Windows I don't see the drop-down control: ![image](https://user-images.githubusercontent.com/1108513/236449575-efdfbb11-b61c-4c69-a6d8-8a5537023969.png)
> 
> Do you still have the Dropdown the missing?

Am onsite right now, without access to MS-Windows. Will check tomorrow morning.

IanMayo commented:
> Do you still have the Dropdown the missing?

Kind-of.  There is no "down-arrow" to act as a clue that there is a drop-down button. But, if I click in that area I do get the `Video` // `PPTX` popup menu.

IanMayo commented:
Hello @saulhidalgoaular - I've just had a play. A couple of items of feedback:
1. When I pressed "Stop recording" a file selector dialog opened. But, I saw that the time was still moving forward. Capture only ended once the output file was selected.  We should pause the stepper on "Stop", and _then_ do the file management stuff.
2. It produced a `.wmv`  of zero size. The issue is probably reported in this stack trace.  Is it due to how the parameters are provided in MS-Windows? 
![image](https://github.com/debrief/debrief/assets/1108513/5bc65f15-a8fc-4397-a51e-48bb78d1eac5)

saulhidalgoaular commented:
> Hello @saulhidalgoaular - I've just had a play. A couple of items of feedback:
> 
> 1. When I pressed "Stop recording" a file selector dialog opened. But, I saw that the time was still moving forward. Capture only ended once the output file was selected.  We should pause the stepper on "Stop", and _then_ do the file management stuff.
> 2. It produced a `.wmv`  of zero size. The issue is probably reported in this stack trace.  Is it due to how the parameters are provided in MS-Windows?
>    ![image](https://private-user-images.githubusercontent.com/1108513/319513432-5bc65f15-a8fc-4397-a51e-48bb78d1eac5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIyMjI3MTksIm5iZiI6MTcxMjIyMjQxOSwicGF0aCI6Ii8xMTA4NTEzLzMxOTUxMzQzMi01YmM2NWYxNS1hOGZjLTQzOTctYTUxZS00OGJiNzhkMWVhYzUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MDRUMDkyMDE5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmI2Nzg0ZWQxODNlZmVkOGUxZGZlNjg5N2FmMzFlNzVkNTMzNGRhNzkwY2Q4NzFlYTk5MDg1N2FlN2JjZWQyZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.UQ8qh51UUK0e7xYtasX_PLzRc1GtM1RvA41vOSiHB4c)

Have you configured ffmpeg? There is a setting section in the Preference for it. It looks like it is missing.

![image](https://github.com/debrief/debrief/assets/11049943/ac305a27-72e1-45b4-90ed-da7002f649e0)

IanMayo commented:
> Have you configured ffmpeg? There is a setting section in the Preference for it. It looks like it is missing.

Yes, that's totally the issue :-)

Is it an option for us to include ffmpeg?  Oh, I see it's quite large.  In that case yes - it's better to have it configurable. Let me install/test.

saulhidalgoaular committed:
Fixing comment 1 from
https://github.com/debrief/debrief/pull/5167#issuecomment-2036582876

IanMayo commented:
@saulhidalgoaular - the _missing-ffmpeg_ issue is probably going to arise a lot.  Can we run a check for the existence of the file in specified in the preference, and throw a suitable warning if the property is empty, or the file doens't exist?

IanMayo commented on org.mwc.cmap.TimeController/src/org/mwc/cmap/TimeController/views/TimeController.java:
```diff
@@ -610,7 +611,13 @@ else if (DEBRIEF_S_WINDOW.equals(preferenceStore.getString(
+      else if (DEBRIEF_PLOT_WINDOW.equals(preferenceStore.getString(
+          P_SCREEN_AREA)))
+      {
+        
```
@saulhidalgoaular - in this `if` structure, should it be retrieving the viewport of the current editor?  Oh, and it shouldn't include the tabs at the top. So, it's the content pane of the current editor, not the whole of the current editor - please.

**Update:** I just tried selecting the new option, but it still records the whole Debrief window.

saulhidalgoaular replied:
Correct. It is still not implemented. I need to find out how to calculate the viewport area. I pushed it partially to fix the bug mentioned in the comment above. 😢

IanMayo replied:
>  I need to find out how to calculate the viewport area.

I think it's something like this:

https://stackoverflow.com/a/11780445/92441

saulhidalgoaular committed:
FFMPEG Validation

saulhidalgoaular committed:
Typo in message box

saulhidalgoaular committed:
Ffmpeg path enable/disable for preference page

saulhidalgoaular committed:
Plot Windows Recording Implemented
